--- vmblock-only/shared/vm_assert.h	2014-04-15 00:41:41.000000000 +0300
+++ vmblock-only/shared/vm_assert.h	2014-04-23 12:44:49.977954180 +0300
@@ -256,6 +256,7 @@ void WarningThrottled(uint32 *count, con
                                        __FILE__, __LINE__, __FUNCTION__, \
                                        _fix))
 #else
+   #undef DEPRECATED /* in <linux/printk.h> since 3.14.0 */
    #define DEPRECATED(_fix) do {} while (0)
 #endif
 
--- vmnet-only/vm_assert.h	2014-04-15 03:06:21.000000000 +0300
+++ vmnet-only/vm_assert.h	2014-04-23 12:44:50.051954176 +0300
@@ -256,6 +256,7 @@ void WarningThrottled(uint32 *count, con
                                        __FILE__, __LINE__, __FUNCTION__, \
                                        _fix))
 #else
+   #undef DEPRECATED /* in <linux/printk.h> since 3.14.0 */
    #define DEPRECATED(_fix) do {} while (0)
 #endif
 
--- vmmon-only/include/vm_assert.h	2014-04-15 03:06:20.000000000 +0300
+++ vmmon-only/include/vm_assert.h	2014-04-23 12:44:50.036954177 +0300
@@ -256,6 +256,7 @@ void WarningThrottled(uint32 *count, con
                                        __FILE__, __LINE__, __FUNCTION__, \
                                        _fix))
 #else
+   #undef DEPRECATED /* in <linux/printk.h> since 3.14.0 */
    #define DEPRECATED(_fix) do {} while (0)
 #endif
 
--- vmci-only/shared/vm_assert.h	2014-04-15 00:41:41.000000000 +0300
+++ vmci-only/shared/vm_assert.h	2014-04-23 12:44:50.021954178 +0300
@@ -256,6 +256,7 @@ void WarningThrottled(uint32 *count, con
                                        __FILE__, __LINE__, __FUNCTION__, \
                                        _fix))
 #else
+   #undef DEPRECATED /* in <linux/printk.h> since 3.14.0 */
    #define DEPRECATED(_fix) do {} while (0)
 #endif
 
